property-scalaz/src/main/scala/testz/property.scala

```diff
@@ -83,44 +84,47 @@ object property {
 
   def exhaustive[F[_]: Applicative, I]
                 (in: List[(() => I)])
-                (testGenerator: I => F[List[TestError]])
-                : F[List[TestError]] = {
-    Traverse[List].traverseM(in)(i => testGenerator(i()))
+                (testGenerator: I => F[TestResult])
+                : F[TestResult] = {
+    in.foldLeft(Success().point[F])((b, a) => Applicative[F].apply2(b, testGenerator(a()))(TestResult.combine))
   }
 
   def exhaustiveU[F[_]: BindRec, I]
                  (in: Unfold[F, I])
-                 (testGenerator: I => F[List[TestError]])
-                 (implicit F: Monad[F]): F[List[TestError]] = {
+                 (testGenerator: I => F[TestResult])
+                 (implicit F: Monad[F]): F[TestResult] = {
     F.join(exhaustiveUR[F, F, I](in)(testGenerator))
   }
 
   def runTests[F[_]: Applicative, G[_]: Applicative, I](
-    testGenerator: I => F[List[TestError]]
-  ): Fold[G, I, F[List[TestError]]] =
-    new Fold[G, I, F[List[TestError]]] {
-      type S = List[F[List[TestError]]]
+    testGenerator: I => F[TestResult]
+  ): Fold[G, I, F[TestResult]] =
+    new Fold[G, I, F[TestResult]] {
+      type S = List[F[TestResult]]
       val start = Nil
-      def step(s: List[F[List[TestError]]], i: I): G[List[F[List[TestError]]]] =
+      def step(s: List[F[TestResult]], i: I): G[List[F[TestResult]]] =
         (testGenerator(i) :: s).pure[G]
-      def end(s: List[F[List[TestError]]]): G[F[List[TestError]]] =
-        s.reverse.traverseM(f => f).pure[G]
+      def end(s: List[F[TestResult]]): G[F[TestResult]] =
+        s.reverse.foldLeft(Success().point[F])(
+          (b, a) => Applicative[F].apply2(b, a)(TestResult.combine)
+        ).pure[G]
```

Comment:
Looks like you never use `G.combine`. Is `G` necessary?

Reply:
Why not `Fold[Id, I, F[TestResult]]`?

Reply:
Good point.

Reply:
This is because otherwise I would be using a `Fold[Id, I, F[TestResult]]` then passing `zapFold` a version which has been converted `Id ~> G` via a combinator, which means megamorphism.